### PR TITLE
Ex4 5 solution

### PR DESCRIPTION
**Pull request by soraisr**

### CODE REVIEW

File: shape/RectByPoints.java [hunk 1]
```diff
@@ -4,10 +4,7 @@
    private Rect rect; // This is called Delegation
 
    public RectByPoints(double[] a, double[] b) {
-	  if (a.length==2 && b.length==2)
 		  this.rect = new Rect( Math.abs(a[1]-a[0]),  Math.abs(b[1]-b[0])); // this should be implemented for real some time :)
```
Comment (aviadbd) [inline shape/RectByPoints.java:7]: Im pretty sure this is not how it works geometrically :)

File: shape/RectByPoints.java [hunk 2]
```diff
@@ -4,10 +4,7 @@
    private Rect rect; // This is called Delegation
 
    public RectByPoints(double[] a, double[] b) {
-	  if (a.length==2 && b.length==2)
```
Comment (aviadbd) [inline shape/RectByPoints.java]: I think some input sanity checks should've been made in the ctor

File: shape/Rect.java [hunk 3]
```diff
@@ -4,7 +4,7 @@
 	private final double height;
 	private final double width;
 
-	Rect(double a , double b){
+	public Rect(double a , double b){
 		this.height = a ;
 		this.width = b ;
```
Comment (aviadbd) [inline shape/Rect.java]: input checks?

File: factory/TriangleFactory.java [hunk 4]
```diff
@@ -1,8 +1,23 @@
...
-    return new Rect(Double.parseDouble(params[0]), Double.parseDouble(params[1]), Double.parseDouble(params[2]);
+  public Shape create(String... params) {
+	  int n = params.length;
+	  if (n == 3)
+	  {
+		  double a = Double.parseDouble(params[0]) ;
+		  double b = Double.parseDouble(params[1]) ;
+		  double c = Double.parseDouble(params[2]) ;
+		  if (a > 0 && c > 0 && b > 0 && a + b > c && a + c > b && c + b > a) 
+			  return new Triangle(a, b, c);
```
Comment (aviadbd) [inline factory/TriangleFactory.java]: I actually agree with your previous comment about checks in the shape classes.

Remember: the shape classes should have the business logic of the shapes, so they need to sanity check their input; however, the input from the user (making sure the string is a double etc) should be done in the factories. Its separation of responsibilities.

File: factory/RectFactory.java [hunk 5]
```diff
@@ -1,8 +1,48 @@
...
+import exceptions.NotEnoughParamException;
+import shape.*;
+
 
 public class RectFactory implements ShapeFactory {
-  Shape create(String... params) {
-    // TODO: Do some parameter checking here and throw appropriate exceptions
-    return new Rect(Double.parseDouble(params[0]), Double.parseDouble(params[1]));
+  public Shape create(String... params) {
+	  int n = params.length ;
```
Comment (aviadbd) [inline factory/RectFactory.java:12]: see my general comment in trianglefactory.

File: factory/CircleFactory.java [hunk 6]
```diff
@@ -1,8 +1,20 @@
...
+import exceptions.*;
 
-public class RectFactory implements ShapeFactory {
-  Shape create(String... params) {
-    // TODO: Do some parameter checking here and throw appropriate exceptions
-    return new Rect(Double.parseDouble(params[0]));
+
+public class CircleFactory implements ShapeFactory {
+  public Shape create(String... params) {
+	  int n = params.length;
```
Comment (aviadbd) [inline factory/CircleFactory.java:8]: see my general comment in trianglefactory.

File: shape/RectByPoints.java (hunk 1 above)
Comment (soraisr) [reply shape/RectByPoints.java:7]: Rect is well defined with 2 points and parallel to the x-axis and y-axis.
The length of the sides is equal to the differences between x and y coordinates.
How to define it in another way? What do you have in mind?

File: shape/Circle.java [hunk 7]
```diff
@@ -1,15 +1,20 @@
 package shape;
 
+import exceptions.NegativeLengthException;
+
 public class Circle implements Shape {
-	private final double radius;
+	protected double radius;
```
Comment (aviadbd) [inline shape/Circle.java:6]: why changed to protected?

File: shape/Circle.java [hunk 8]
```diff
@@ -1,15 +1,20 @@
...
 
+import exceptions.NegativeLengthException;
+
 public class Circle implements Shape {
-	private final double radius;
+	protected double radius;
 	public Circle(double radius){
 		this.radius = radius ;
+		if (radius < 0)
+			throw new NegativeLengthException(" required Radius > 0 ");
```
Comment (aviadbd) [inline shape/Circle.java:10]: always have scope ( brackets ), i.e.:

```suggestion
		if (radius < 0) {
			throw new NegativeLengthException(" required Radius > 0 "); 
		}

```

**Pull request merged**